File: egroot/src/pkg/sync/sync.go
// Package sync provides synchronisation primitives.
package sync

// Barrier is a full compiler memory barrier. All compiler optimizations
// that can reorder explicit memory accesses (at source code level) can't
// cross this barrier.
func Barrier()

// Memory is a hardware full memory barrier. All explicit memory accesses
// (at instruction level) before Memory should be finished before first
// subsequent explicit memory access. Memory implies Barrier.
func Memory()

// Sync is similar to Memory but stronger: it ensures that all explicit memory
// accesses before Sync should be finished before first subsequent instruction.
// Sync implies Barrier.
func Sync()

File: egroot/src/pkg/sync/sync.c
// +build !cortexm3,!cortexm4,!cortexm4f

static int mem;

// This isn't formally correct, but function call (never inlined), followed by
// hardware full memory barrier, followed by R/W memory access, followed by
// return from function, should be enough on all real platform. 
void sync_Sync() {
	__sync_synchronize();
	mem++
}